bot/services/paginate_service.py

```diff
@@ -0,0 +1,119 @@
+from bot.data.message_repository import MessageRepository
+from bot.services.base_service import BaseService
+from bot.messaging.events import Events
+from bot.messaging import messenger
+
+log = logging.getLogger(__name__)
+
+@dataclass
+class Message:
+    embed_name: str
+    field_title: str
+    pages: t.List[str]
+    _curr_page: int
+    author: int
+
+    @property
+    def curr_page(self):
+        return self._curr_page
+
+    @curr_page.setter
+    def curr_page(self, page: int):
+        self._curr_page = page
+
+    @property
+    def curr_content(self):
+        return self.pages[self._curr_page]
+
+class PaginateService(BaseService):
+    """
+    This service allows for messages sent by the bot to be paginated
+    """
+
+    def __init__(self, *, bot):
+        super().__init__(bot)
+        self.messages = {}
+        self.reactions = ["⏮️","⬅️","➡️","⏭️"]
+
+    # Called When a cog would like to be able to paginate a message
+    @BaseService.Listener(Events.on_set_pageable)
+    async def set_message_pageable(self, *,
+                                    embed_name: str,
+                                    field_title: str,
+                                    pages: t.List[str],
+                                    author: discord.Member = None,
+                                    channel: discord.TextChannel,
+                                    timeout: int = 60):
+
+        if not isinstance(pages , t.List):
+            pages = [pages]
+
+        embed = discord.Embed(title= embed_name, color= Colors.ClemsonOrange)
+        # set the first page of the embed
+        embed.add_field(name= field_title, value= pages[0])
+        embed.set_footer(text=f'Page 1 of {len(pages)}')
+        msg = await channel.send(embed= embed)
+        await self.bot.messenger.publish(Events.on_set_deletable, msg=msg, author=msg.author)
+
+        # stores the message info
+        message = Message(embed_name, field_title, pages, 0, author.id if author else None)
+        self.messages[msg.id] = message
+
+        # add every emoji from the reaction list
+        for reaction in self.reactions:
+            await msg.add_reaction(reaction)
+
+        if timeout:
+            await asyncio.sleep(timeout) 
+            try:
+                for reaction in self.reactions:
+                    await msg.clear_reaction(reaction)
+                del self.messages[msg.id]
+            except:
+                pass
+            finally:
+                log.info(f'Message: {msg.id} timed out as pageable')
+
+    @BaseService.Listener(Events.on_reaction_add)
+    async def change_page(self, reaction: discord.Reaction, user: t.Union[discord.User, discord.Member]):
+        msg = self.messages[reaction.message.id]
+
+        # check if emoji matches and user has perm to change page
+        if (reaction.emoji not in self.reactions 
+             or reaction.message.id not in self.messages 
+             or not user.guild_permissions.administrator 
```

Comment:
this part right here means that you cant scroll unless you are an admin, this should be 
```suggestion
         if not user.guild_permissions.administrator and not user.id == msg.author:
             return
```

Reply:
i like to keep my control flow as flat as possible, and a big way to do that is isolate return points that have X success cases and only 1 or 2 failure points. Instead of checking for each success case, you can instead invert the logic and merely check the 1 failure point and return on that

Reply:
In this case there are 3 success and 1 failure, with the only time it should fail being when a user is not an admin AND not the author of the embed